pkg/apis/core/v1alpha1/helper/shootstate_list.go

```diff
@@ -36,10 +36,9 @@ func (e *ExtensionResourceStateList) Get(kind string, name, purpose *string) *ga
 
 // Delete removes an ExtensionResourceState from the list by kind, name and purpose
 func (e *ExtensionResourceStateList) Delete(kind string, name, purpose *string) {
-	for i, obj := range *e {
-		if matchesExtensionResourceState(&obj, kind, name, purpose) {
+	for i := len(*e) - 1; i >= 0; i-- {
+		if matchesExtensionResourceState(&(*e)[i], kind, name, purpose) {
 			*e = append((*e)[:i], (*e)[i+1:]...)
-			return
```

Comment:
There is one issue if the return is not removed - it will remove only the first occurrence of an Extension matching kind,name and purpose and keep the rest. Most probably it is used in a way that duplication do not happen, but I have not checked for that.

Reply:
Yep, the list is meant to contain only uniquely identifiable items. We didn't chose a map because lists are better from API perspective. You can check that the `Upsert` method looks for an already existing item and changes that if it finds something, instead of inserting a duplicate.

Reply:
I have re-added the return. Also added it in the other `Delete` methods.